[4.0] Ravendb 10538

aviv86 opened the pull request.

aviv86 committed:
RavenDB-10538 : facets should use parameters (client side)

aviv86 committed:
RavenDB-10538 : facets should use parameters (server side)

aviv86 committed:
RavenDB-10538 : fix old facet tests, add new test cases

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -55,23 +56,29 @@ internal override RangeFacet AsRangeFacet()
         }
         
-        public static string Parse(string prefix, LambdaExpression expr)
+        public static string Parse(string prefix, LambdaExpression expr, Parameters parameters)
```
This will break a binary and source-level compatibility

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -55,23 +56,29 @@ internal override RangeFacet AsRangeFacet()
         }
 
-        public static string Parse(Expression<Func<T, bool>> expr)
+        public static string Parse(Expression<Func<T, bool>> expr, Parameters parameters = null)
```
This will break a binary and source-level compatibility

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/FacetBase.cs:
```diff
@@ -17,6 +17,8 @@ protected FacetBase()
 
         public Dictionary<FacetAggregation, string> Aggregations { get; set; }
 
+        public Parameters FacetParameters { get; set; }
```
Wouldn't it be better to delay the conversion of Facet (either `Facet`, `Facet<T>`, `RangeFacet` or `RangeFacet<T>`) to the last possible point (where FacetToken is created). You would not have to drag the parameters along the way, then rewrite them to query parameters, rewrite the expressions, etc. This would cause much less trouble.

ppekrol replied:
Is this needed now? I do not see any usages.

aviv86 committed:
RavenDB-10538:
range facet: parse the Range-Expressions at the last possible point (where FacetToken is created),
in order to avoid rewriting the facet-parameters to query parameters and rewriting the ranges.

aviv86 committed:
RavenDB-10538 : fix tests

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -20,28 +21,12 @@ public RangeFacet()
         /// </summary>
-        public List<Expression<Func<T, bool>>> Ranges { get; set; }
+        [JsonIgnore]
+        public List<LambdaExpression> RangeExpressions { get; set; }
```
This should be:
a) internal
b) in `RangeFacet<T>`, not RangeFacet

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -53,25 +38,7 @@ internal override RangeFacet AsRangeFacet()
-        }
-        
-        public static string Parse(string prefix, LambdaExpression expr)
+        public static string Parse(string prefix, LambdaExpression expr, Func<object, string> addQueryParameter)
```
This will break a binary compatibility. You need to have two methods here:
```
public static string Parse(string prefix, LambdaExpression expr)

public static string Parse(string prefix, LambdaExpression expr, Func<object, string> addQueryParameter)
```

The first one needs to pass `null` to the second one.

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -243,7 +210,7 @@ private static object ParseUnaryExpression(UnaryExpression expression)
         }
 
-        private static object TryInvokeLambda(Expression expression)
+        public static object TryInvokeLambda(Expression expression)
```
Why public?

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/RangeFacet.cs:
```diff
@@ -269,17 +236,25 @@ private static string GetStringRepresentation(string fieldName, ExpressionType l
         {
-            switch (value.GetType().FullName)
+
+            if (addQueryParameter != null &&
```
What about other types here? Shouldn't this be inside the switch?

ppekrol commented on src/Raven.Client/Documents/Queries/Facets/FacetSetup.cs:
```diff
@@ -17,7 +17,21 @@ public class FacetSetup
+            {
+                foreach (var rangeFacet in value)
+                {
+                    rangeFacet.RangeExpressions = null;
```
Is this needed? I saw the [JsonIgnore] attribute on that property.

ppekrol commented:
@aviv86 check this approach: https://github.com/ravendb/ravendb/commit/672c154a07ddfc737a6bff97ed814e1b7c7033e2

It covers all cases except creating `RangeFacet<T>` inside a `List<RangeFacet>` due to implicit conversion that exists between those two types. I do not think that this should be a major problem since most of people will use either SetupDoc or FacetBuilder (and it will work inside `List<FacetBase>` or `List<RangeFacet<T>>` anyway)

We still need to handle other types when parsing ranges, haven't fixed that :)
